src/source/Ice/TurnConnection.c

```diff
@@ -1048,10 +1057,9 @@ STATUS turnConnectionStepState(PTurnConnection pTurnConnection)
                 }
 
                 CHK_STATUS(turnConnectionFreePreAllocatedPackets(pTurnConnection));
-
+                CHK_STATUS(socketConnectionClosed(pTurnConnection->pControlChannel));
```

Comment:
I am wondering if you still need to set the shutdownComplete atomic in case of failure anyway?

Reply:
I added logic that when failure happens, turn will go to cleanup state, then go to failed state. Also added unit test for that.